src/auth/auth.controller.ts

```diff
@@ -89,4 +90,29 @@ export class AuthController {
       accessToken: await this.authService.issueAccessToken(token),
     }
   }
+
+  private createOauthState(returnUrl: string, overrideBackendUrl?: string) {
+    const state: OauthStatePayload = {
+      returnUrl: returnUrl || this.configService.get('backendPublicUrl'),
+    }
+    if (overrideBackendUrl) {
+      state.overrideBackendUrl = overrideBackendUrl
+    }
+    return state
+  }
+
+  private validateOauthState(state: OauthStatePayload) {
+    const backendUrl = this.configService.get('backendPublicUrl')
+    if (this.configService.get('env') === 'development') return
+    if (state.overrideBackendUrl) {
+      throw new BadRequestException(
+        `overriding backendUrl is not allowed in this environment`
+      )
+    }
+    if (!state.returnUrl.startsWith(backendUrl)) {
+      throw new BadRequestException(
+        `returnUrl must be the same origin as ${backendUrl}`
+      )
+    }
```

Comment:
Oh, This is not the correct way to check for origin isn't it?

Reply:
Why not? Our `backendUrl` is an origin

Reply:
In https://github.com/thinc-org/cugetreg-infrastructure/blob/master/src/production/backend/.env.production, I could just do `https://cugetreg.com.malicious.domain/...` to bypass this check.

Reply:
how do u propose we check for origin then?

Reply:
I think the best low-cost solution is to just parse the url, and verify that origin and path are in the allowlist.

Reply:
Maybe put this in the issue backlog. I can see that this bug is not introduced in this pr.

Reply:
> how do u propose we check for origin then?

See ce3c407